macaroonbakery/bakery.py

```diff
@@ -0,0 +1,280 @@
+            except KeyError:
+                raise DischargeError(
+                    'macaroon not found in unauthorized response {}'.format(
+                        response.content))
+            macaroon = deserialize(serialized_macaroon)
+            discharges = self._discharge(macaroon)
+            encoded_discharges = map(serialize_macaroon_string, discharges)
+            self._macaroons = base64.urlsafe_b64encode(
+                '[' + ','.join(encoded_discharges) + ']')
+            return self.get(url, **kwargs)
+
+        return response
+
+    def _discharge(self, macaroon):
+        '''Discharge a macaroon.
+
+        @param the macaroon object to discharge.
+        @returns the discharged macaroons as an array of macaroon object.
+        @raise DischargeError when an error occurs during the discharge
+            process.
+        '''
+        discharges = [macaroon]
+        first_party_location = macaroon.location
+        self._discharge_caveats(macaroon, discharges, first_party_location,
+                                macaroon)
+        return discharges
+
+    def _discharge_caveats(self, macaroon, discharges, first_party_location,
+                           primary_macaroon):
+        '''Gathers discharge macaroons for all the third party caveats
+           for the macaroon passed in.
+
+        @param macaroon the macaroon to discharge.
+        @param discharges the list of discharged macaroons.
+        @param first_party_location the location of the first party.
+        @param primary_macaroon used for signature of the discharge macaroon.
+        @raise DischargeError when an error occurs during the discharge
+            process.
+        '''
+        caveats = macaroon.third_party_caveats()
+        for caveat in caveats:
+            dm = self._get_discharge(first_party_location,
+                                     caveat[0],
+                                     caveat[1])
+            dm = primary_macaroon.prepare_for_request(dm)
+            discharges.append(dm)
+            self._discharge_caveats(dm, discharges, first_party_location,
+                                    primary_macaroon)
+
+    def _get_discharge(self, location, third_party_location, condition):
+        ''' Get the discharge macaroon from the third party location.
+
+        @param location the origin.
+        @param third_party_location where to get a discharge from.
+        @param condition associated  to the discharged macaroon.
+        @return a discharged macaroon.
+        @raise DischargeError when an error occurs during the discharge
+            process.
+        '''
+        headers = {'Bakery-Protocol-Version': 1}
+        if self._discharge_token is not None:
+            headers['Macaroons'] = self._discharge_token
+        payload = {'id': condition, 'location': location}
+
+        try:
+            response = requests.post(third_party_location + '/discharge',
```

Comment:
I think this should recursively call the bakery service rather than invoking requests.post directly, so that it will automatically acquire authorization if necessary. Some services already rely on that behaviour.